File: new/python/74_search_a_2d_matrix_binary_search.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # idea: binary search
        # do a first while loop to find out what might be the correct row (doesn't guarentee that
        # we have the value in that row, atleast we eliminate other rows).
        # then do another while loop to find of the row we found out, if the value exists, which
        # column (cell) does it exist.
        ROWS = len(matrix)
        COLS = len(matrix[0])

        bottom_row = 0
        top_row = ROWS - 1

        while bottom_row <= top_row:
            candidate_row = bottom_row + (top_row - bottom_row) // 2
            first_element_of_current_row = matrix[candidate_row][0]
            last_element_of_current_row = matrix[candidate_row][COLS - 1]
            if target < first_element_of_current_row:
                top_row = candidate_row - 1
            elif target > last_element_of_current_row:
                bottom_row = candidate_row + 1
            else:
                break

        if not (bottom_row <= top_row):
            return False

        candidate_row = bottom_row + (top_row - bottom_row) // 2

        left_col = 0
        right_col = COLS - 1

        while left_col <= right_col:
            candidate_col = left_col + (right_col - left_col) // 2
            current_val = matrix[candidate_row][candidate_col]
            if current_val < target:
                left_col = candidate_col + 1
            elif current_val > target:
                right_col = candidate_col - 1
            else:
                return True

        return False
        # time: O(logm + logn) => O(log(m * n))
        # space: O(1)

File: new/cpp/74_search_a_2d_matrix_binary_search.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // idea: binary search
        // first while loop to find the correct row (binary search the correct row)
        // second while loop to find the correct column (binary search the correct column)
        int ROWS = matrix.size();
        int COLS = matrix[0].size();

        int br = 0; // bottom row means 0th row
        int tr = ROWS - 1; // top row means from the last row and below

        while (br <= tr) {
            // this is the loop where we find the row which could potentially have our target
            int mr = br + (tr - br) / 2; 
            // middle row is used for our binary search to adjust
            // values of br, and tr.
            int firstElementOfMr = matrix[mr][0]; // first element of middle row
            int lastElementOfMr = matrix[mr][COLS - 1]; // last element of middle row

            if (target < firstElementOfMr) { // if the target is less than first element of mr
                tr--;
            } else if (target > lastElementOfMr) { // if target > last element of middle row
                br++;
            } else {
                break; // because we found potential row where our value could exist
                // because the target is not less than the first elem of middle row nor
                // greater than the last elem of middle row
            }
        }

        if (!(br <= tr)) {
            return false;
        }


        int mr = br + (tr - br) / 2;  // we declare again because the previous declaration
        // is inside the while loop and it is not available in the global scope. 

        int lc = 0;
        int rc = COLS - 1;

        while (lc <= rc) {
            // second while loop to find out if the target is present in our middle row
            // and which column
            int mc = lc + (rc - lc) / 2;
            int currentVal = matrix[mr][mc];
            if (target < currentVal) {
                rc = mc - 1;
            } else if (target > currentVal) {
                lc = mc + 1;
            } else {
                return true;
            }
        }

        return false;
    }
    // time: O(log(m) + log(n)) => O(log(m * n))
    // space: O(1)
};
